File: PROGRAMMERS/체육복.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>
#include <cstring>
using namespace std;

#define endl '\n'

typedef long long ll;
typedef pair<int, int> pii;
typedef pair<ll, ll> pll;

int solution(int n, vector<int> lost, vector<int> reserve)
{
  vector<int> answer(n,1);
  for (int i = 0; i < (int)lost.size(); ++i) {
    answer[lost[i]-1]--;
  }
  for (int i = 0; i < (int)reserve.size(); ++i) {
    answer[reserve[i]-1]++;
  }
  for (int i = 0; i < (int)answer.size(); ++i) {
    if (!i) {
      if (answer[i] == 0) {
        if (answer[i + 1] == 2) {
          answer[i] += 1;
          answer[i + 1] -= 1;
        }
      }
    }
    else if (i == (int)answer.size() - 1) {
      if (answer[i] == 0) {
        if (answer[i - 1] == 2) {
          answer[i] += 1;
          answer[i - 1] -= 1;
        }
      }
    }
    else {
      if (answer[i] == 0) {
        if (answer[i - 1] == 2) {
          answer[i - 1] -= 1;
          answer[i] += 1;
        }
        else if (answer[i + 1] == 2) {
          answer[i + 1] -= 1;
          answer[i] += 1;
        }
      }
    }
  }
  int ret = 0;
  for (int i = 0; i < (int)answer.size(); ++i) {
    if (answer[i]) {
      ret++;
    }
  }  
  return ret;
}

int main()
{
  ios_base::sync_with_stdio(false);
  cout.tie(nullptr); cin.tie(nullptr);

  int n = 5;
  vector<int> lost = { 1,3 };
  vector<int> reserve = { 2,3,4 };

  int ret = solution(n,lost,reserve);
  cout << ret << endl;
  return 0;
}

File: PS_Python/체육복.py
def solution(n, lost, reserve):

    lost = list(map(lambda x: x - 1, lost))
    reserve = list(map(lambda x: x - 1, reserve))
    answer = 0

    dress = [1] * n

    for l in lost:
        dress[l] -= 1
    for r in reserve:
        dress[r] += 1

    for idx, d in enumerate(dress):
        if idx == 0:
            if dress[idx] == 0:
                if dress[idx + 1] == 2:
                    dress[idx] += 1
                    dress[idx + 1] -= 1
        elif idx == len(dress) - 1:
            if dress[idx] == 0:
                if dress[idx - 1] == 2:
                    dress[idx] += 1
                    dress[idx - 1] -= 1
        else:
            if dress[idx] == 0:
                if dress[idx + 1] == 2:
                    dress[idx] += 1
                    dress[idx + 1] -= 1
                elif dress[idx - 1] == 2:
                    dress[idx] += 1
                    dress[idx - 1] -= 1

    answer = 0
    for d in dress:
        if d:
            answer += 1
    return answer


def solution2(n, lost, reserve):
    _reserve = [r for r in reserve if r not in lost]
    _lost = [l for l in lost if l not in reserve]
    for r in _reserve:
        f = r - 1
        b = r + 1
        if f in _lost:
            _lost.remove(f)
        elif b in _lost:
            _lost.remove(b)
    return n - len(_lost)


print(solution2(5, [2, 4], [1, 3, 5]))

